js/modules/k6/experimental/tracing/client.go

```diff
@@ -110,41 +108,79 @@ func (c *Client) Request(method string, url goja.Value, args ...goja.Value) (*ht
 	}
 
-	// Produce a trace header in the format defined by the
-	// configured propagator.
+	// Produce a trace header in the format defined by the configured propagator.
 	traceContextHeader, err := c.propagator.Propagate(encodedTraceID)
 	if err != nil {
-		return nil, fmt.Errorf("failed to propagate trace ID; reason: %w", err)
+		return http.Header{}, "", fmt.Errorf("failed to propagate trace ID; reason: %w", err)
 	}
 
+	return traceContextHeader, encodedTraceID, nil
+}
+
+// Request instruments the http module's request function with tracing headers,
+// and ensures the trace_id is emitted as part of the output's data points metadata.
+func (c *Client) Request(method string, url goja.Value, args ...goja.Value) (*httpmodule.Response, error) {
+	var result *httpmodule.Response
+	var err error
+	err = c.instrumentedCall(func(args ...goja.Value) error {
+		result, err = c.requestFunc(method, url, args...)
+		return err
+	}, args...)
+
+	if err != nil {
+		return nil, err
+	}
+	return result, nil
+}
+
+// AsyncRequest instruments the http module's asyncRequest function with tracing headers,
+// and ensures the trace_id is emitted as part of the output's data points metadata.
+func (c *Client) AsyncRequest(method string, url goja.Value, args ...goja.Value) (*goja.Promise, error) {
+	var result *goja.Promise
+	var err error
+	err = c.instrumentedCall(func(args ...goja.Value) error {
+		result, err = c.asyncRequestFunc(method, url, args...)
+		return err
+	}, args...)
+
+	if err != nil {
+		return nil, err
+	}
+	return result, nil
+}
+
+func (c *Client) instrumentedCall(call func(args ...goja.Value) error, args ...goja.Value) error {
+	if len(args) == 0 {
+		args = []goja.Value{goja.Null()}
+	}
+
+	traceContextHeader, encodedTraceID, err := c.generateTraceContext()
+	if err != nil {
+		return err
+	}
 	// update the `params` argument with the trace context header
 	// so that it can be used by the http module's request function.
 	args, err = c.instrumentArguments(traceContextHeader, args...)
 	if err != nil {
-		return nil, fmt.Errorf("failed to instrument request arguments; reason: %w", err)
+		return fmt.Errorf("failed to instrument request arguments; reason: %w", err)
 	}
 
 	// Add the trace ID to the VU's state, so that it can be
 	// used in the metrics emitted by the HTTP module.
 	c.vu.State().Tags.Modify(func(t *metrics.TagsAndMeta) {
 		t.SetMetadata(metadataTraceIDKeyName, encodedTraceID)
 	})
-
-	response, err := c.requestFunc(method, url, args...)
-	if err != nil {
-		return nil, err
-	}
-
-	// Remove the trace ID from the VU's state, so that it doesn't
-	// leak into other requests.
-	c.vu.State().Tags.Modify(func(t *metrics.TagsAndMeta) {
-		t.DeleteMetadata(metadataTraceIDKeyName)
-	})
-
-	return response, nil
+	// Remove the trace ID from the VU's state, so that it doesn't leak into other requests.
+	defer func() {
+		c.vu.State().Tags.Modify(func(t *metrics.TagsAndMeta) {
+			t.DeleteMetadata(metadataTraceIDKeyName)
```

Comment:
If I remove this line nothing fails between the HTTP and tracing module tests

Reply:
I do think this should be addressed separately. After taking a look this is not new I just refactor the code - if anything my code is less likely to break :P. But writing a test for this seems a bit more involved than I would prefer for this PR

Reply:
I did open an issue with it https://github.com/grafana/k6/issues/2889